### PR TITLE
Retry Sendto() for IPv6 NA for tentative IPv6 addresses and EADDRNOTAVAIL

### PR DESCRIPTION
**Pull request by thom311**
When kernel initially adds an IPv6 address, it is still tentative and
pending DuplicateAddressDetection. Trying to use the address results in
EADDRNOTAVAIL error.

Detect that error, and retry for up to 3 seconds to send the message.

Unfortunately, DAD can easily take more than a second, so this change
delays the completion of the CNI binary quite a bit.

---

@SchSeba  Follow-up for https://github.com/k8snetworkplumbingwg/sriov-cni/pull/306

This seems to happen with Openshift 4.17, which makes me wonder whether previous versions had ~opportunistic~optimistic DAD enabled. If so, why was it disabled?

### CODE REVIEW

File: pkg/utils/packet.go
Comment (wizhaoredhat) [inline pkg/utils/packet.go:215]: This looks awkward. Why do we need to do this?
I don't quite understand why we need to loop twice?

Comment (thom311) [reply pkg/utils/packet.go:215]: to apply IPv4 addresses first (during the first iteration).

Comment (thom311) [reply pkg/utils/packet.go:215]: to quote the commit message:
```
Announce fir IPv4 addresses (GARP) before IPv6 NA

To announce IPv6 addresses, SendUnsolicitedNeighborAdvertisement()
possibly needs to wait for tentative IPv6 addresses to become usable
(DAD). This can take more than a second.

Change the loop over all addresses, to iterate them twice, and on the
first run announce IPv4 addresses. IPv4 addresses  don't need to be
affected by this waiting.

Note that utils.AnnounceIPs() is called towards the end of whatever the
command invocation is doing. This reordering, ensures that we
potentially announce IPv4 addresses earlier.
```
Is that unclear?

Comment (wizhaoredhat) [reply pkg/utils/packet.go:215]: Ah that's the problem. Yes the commit message makes it clear. But in code, as a reader, you don't really know what is going on. It may not be easy to look back in history esp. when the code gets refactored or moved.
This change is a hopeful optimization where there is a chance when doing the GARP for IPv4, it gives enough time for IPv6 DAD to finish.
This optimization comes at a cost of readability of the code. Maybe it would be better to make it more explicit. By having 2 for loops instead of 1 nested for loop.

Comment (thom311) [reply pkg/utils/packet.go]: reworked the branch. Better now?

Comment (wizhaoredhat) [reply pkg/utils/packet.go]: Looks good!

File: pkg/utils/packet.go
Comment (zeeke) [inline pkg/utils/packet.go:174]: This condition looks flipped to me. 

```suggestion
			if time.Now().Before(retryUntil) {
```

Any chance to get a unit test on this function?

Comment (thom311) [reply pkg/utils/packet.go]: This seems hard to unit test.

**Comment by adrianchiris** [comment]
delaying CNI call for a single interface by up to 3 seconds for an optimization mechanism is very problematic in my view.

**Comment by thom311** [comment]
> delaying CNI call for a single interface by up to 3 seconds f

In practice, it seems to be more like 1.8 seconds. The 3 seconds is a maximum (which the patch maybe should reduce to 2 seconds). Still, it's long.

> delaying CNI call for a single interface by up to 3 seconds for an optimization mechanism is very problematic in my view.

Note that the IPv6 addresses are not really usable on the POD anyway. It also means, to start the pod with non-complete IP configuration (e.g. they cannot bind to the IPv6 address). So delaying the startup may be desirable also for those reasons. However, to have this wait implemented via AnnounceIPs() is non-obvious.

> moreover if the first(ipv6) ip consumed these 3 seconds, then other IP6s will not have neighbour advertisment.

I don't understand this. If the first IPv6 was delayed by the full 3 seconds, then indeed, subsequent addresses won't be delayed/retried. But they are still tried at least once.

> having opportunistic DAD enabled seems like the correct approach.
how do you enable that in the system ?

Sorry, I misspoke earlier: `s/opportunistic/optimistic/`

I agree. Something like `echo 1 > /proc/sys/net/ipv6/conf/eth0/optimistic_dad`

**Comment by adrianchiris** [comment]
> I don't understand this. If the first IPv6 was delayed by the full 3 seconds, then indeed, subsequent addresses won't be delayed/retried. But they are still tried at least once.

yes, i re-read the code. you are correct.

**Comment by thom311** [comment]
@adrianchiris branch completely reworked. Now, optimistic_dad is set.

This is a far-reaching change, I am not sure I understand all implications. It does however seem the right thing to me, also because it seems bad to hand control over to the container with IPv6 addresses that are not ready yet.

The retry is gone because it should no longer be necessary. It would however not hurt, because the retry was only done when it was determined to be necessary.

WDYT?

**Comment by wizhaoredhat** [comment]
I really like the opportunistic DAD approach! It looks much better now. Also IPAM should really never give a duplicate IPv6 address when setup properly. As Adrian said, duplicate IPv6 addresses shouldn't occur.

**Comment by SchSeba** [comment]
but we still need to remember the case where a user wants a static IP address on a pod and he restart the pod.

that is a valid use-case

**Comment by thom311** [comment]
> but we still need to remember the case where a user wants a static IP address on a pod and he restart the pod.

I don't follow. Could you please elaborate?

File: pkg/utils/packet.go
Comment (SchSeba) [inline pkg/utils/packet.go:185]: In this case, we only sent the last error.

We may want to append all the errors so we can see in debug mode what ips configuration failed.

File: pkg/utils/packet.go
Comment (SchSeba) [inline pkg/utils/packet.go:194]: same as above

File: pkg/utils/packet.go
Comment (thom311) [reply pkg/utils/packet.go]: is that better?

**Review by SchSeba — APPROVED** [review]
LGTM

File: pkg/utils/utils.go
Comment (adrianchiris) [inline pkg/utils/utils.go:55]: lets not return error if path does not exist.

File: pkg/utils/packet.go
Comment (adrianchiris) [inline pkg/utils/packet.go:179]: nit: can you document this new behaviour in AnnounceIPs docstring ?

**Comment by adrianchiris** [comment]
overall LGTM, once comments addressed we can merge.

File: pkg/utils/utils.go
Comment (thom311) [reply pkg/utils/utils.go:55]: could you elaborate why? If the path does not exist (which is most likely because the interface was just renamed or ipv6 module is not loaded), it's still an error. Granted, the caller ignores the return value anyway.

File: pkg/utils/packet.go
Comment (thom311) [reply pkg/utils/packet.go:183]: done. How is this?

File: pkg/utils/utils.go
Comment (adrianchiris) [reply pkg/utils/utils.go:55]: my though was: if path does not exist there is nothing to enable - so this func should be a no-op.
since ipv6 modules may not me loaded. instead of always trying to write.

lets keep as is.

**Review by adrianchiris — APPROVED** [review]

File: pkg/utils/utils.go
Comment (thom311) [reply pkg/utils/utils.go:55]: Since the caller ignores the error it's kinda moot anyway.

It the caller ever will do something about the error, we know better what they expect. Then we can fine tune the desired error behavior. Which will not be a problem, since this is only internal API that can change.

**Pull request merged**
